### PR TITLE
feat (api): application dao rework & metadata

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/application/dao.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package application
+
+import (
+	"github.com/go-gorp/gorp"
+
+	"database/sql"
+
```
Comment (sguiheux) [inline engine/api/application/dao.go]: wrong place

File: engine/api/application/dao_dependencies.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
+package application
+
+import (
+	"github.com/go-gorp/gorp"
+
+	"database/sql"
+
```
Comment (sguiheux) [inline engine/api/application/dao_dependencies.go]: import

File: engine/api/application/dao_dependencies.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	loadGroups = func(db gorp.SqlExecutor, app *sdk.Application, u *sdk.User) error {
+		if err := LoadGroupByApplication(db, app); err != nil && err != sql.ErrNoRows {
+			return sdk.WrapError(err, "application.loadGroups> Unable to load group permission for application %d", app.ID)
+		}
+		return nil
+	}
+
+	loadPermission = func(db gorp.SqlExecutor, app *sdk.Application, u *sdk.User) error {
+		app.Permission = permission.ApplicationPermission(app.ID, u)
```
Comment (sguiheux) [inline engine/api/application/dao_dependencies.go]: why not void ? or permission.ApplicationPermission can return an error

File: engine/api/application/dao_dependencies.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		return nil
+	}
+
+	loadNotifs = func(db gorp.SqlExecutor, app *sdk.Application, u *sdk.User) error {
+		var err error
+		app.Notifications, err = notification.LoadAllUserNotificationSettings(db, app.ID)
+		if err != nil && err != sql.ErrNoRows {
+			return sdk.WrapError(err, "application.loadNotifs> Unable to load notifications for application %d", app.ID)
+		}
+		return err
```
Comment (sguiheux) [inline engine/api/application/dao_dependencies.go]: return nil or you really want to return sql.ErrNoRows?

File: engine/api/application/gorp_model.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package application
+
+import (
+	"encoding/json"
+
+	"github.com/go-gorp/gorp"
```
Comment (sguiheux) [inline engine/api/application/gorp_model.go]: import

File: engine/api/application/gorp_model.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	if metadataStr.Valid {
+		metadata := sdk.Metadata{}
+		if err := json.Unmarshal([]byte(metadataStr.String), &metadata); err != nil {
+			return err
+		}
+		a.Metadata = metadata
+	}
+
+	pkey, errP := db.SelectStr("select projectkey from project where id = $1", a.ProjectID)
```
Comment (sguiheux) [inline engine/api/application/gorp_model.go]: Load project key  in load application by name

File: engine/api/application_pipeline.go [hunk 7]
```diff
@@ -78,7 +79,9 @@ func attachPipelinesToApplicationHandler(w http.ResponseWriter, r *http.Request,
 		return err
 	}
 
-	app, err := application.LoadApplicationByName(db, key, appName)
+	fmt.Println(c.User)
```
Comment (sguiheux) [inline engine/api/application_pipeline.go]: delete

File: engine/api/application/dao_dependencies.go (hunk 3 above)
Comment (fsamin) [inline engine/api/application/dao_dependencies.go]: permission.ApplicationPermission doesnt return error

File: engine/api/application/gorp_model.go (hunk 5 above)
Comment (fsamin) [inline engine/api/application/gorp_model.go]: fixed

File: engine/api/application/gorp_model.go (hunk 6 above)
Comment (fsamin) [inline engine/api/application/gorp_model.go]: humm not so easy. Try different things without any success. 
Last chance : add it in application table... what do you think about it ?

File: sdk/application.go [hunk 8]
```diff
@@ -160,6 +160,29 @@ func GetApplication(pk, name string, opts ...RequestModifier) (*Application, err
...
+	url := fmt.Sprintf("/project/%s/application/%s", app.ProjectKey, app.Name)
+	data, code, err := Request("PUT", url, data)
+	if err != nil {
+		return err
+	}
+
+	if code != http.StatusOK {
+		return fmt.Errorf("Error [%d]: %s", code, data)
+	}
+	if e := DecodeError(data); e != nil {
```
Comment (yesnault) [inline sdk/application.go]: return DecodeError(data)

**Comment by fsamin** [comment]
all is green @yesnault @sguiheux

**Pull request merged**